src/event-log.c

```diff
@@ -0,0 +1,398 @@
+#include <errno.h>
+#include <glib.h>
+#include <gio/gio.h>
+#include <glib/gstdio.h>
+#if ENABLE_JSON
+#include <json-glib/json-glib.h>
+#endif
+
+#include "context.h"
+#include "event-log.h"
+
+static const gchar *supported_event_types[] = {
+	"all",
+	R_EVENT_LOG_TYPE_BOOT,
+	R_EVENT_LOG_TYPE_INSTALL,
+	R_EVENT_LOG_TYPE_SERVICE,
+	R_EVENT_LOG_TYPE_WRITE_SLOT,
+	R_EVENT_LOG_TYPE_BOOT_SELECTION,
+};
+
+gboolean r_event_log_is_supported_type(const gchar *type)
+{
+	return g_strv_contains(supported_event_types, type);
+}
+
+void r_event_log_message(const gchar *type, const gchar *message, ...)
+{
+	va_list list;
+	g_autofree gchar *formatted = NULL;
+
+	g_return_if_fail(message);
+
+	va_start(list, message);
+	formatted = g_strdup_vprintf(message, list);
+	va_end(list);
+
+	g_log_structured(R_EVENT_LOG_DOMAIN, G_LOG_LEVEL_MESSAGE,
+			"RAUC_EVENT_TYPE", type,
+			"MESSAGE", "%s", formatted);
+}
+
+void r_event_log_free_logger(REventLogger *logger)
+{
+	if (!logger)
+		return;
+
+	g_clear_pointer(&logger->name, g_free);
+	g_clear_pointer(&logger->filename, g_free);
+	g_clear_pointer(&logger->logstream, g_object_unref);
+
+	g_free(logger);
+
+	return;
+}
+
+/**
+ * Log formatter for json output.
+ *
+ * Simply converts all log fields into a json object.
+ *
+ * | {
+ * |   "date": "2023-06-14T21:15:41Z"
+ * |   "MESSAGE" : "Booted into rootfs.0 (A)",
+ * |   "MESSAGE_ID" : "e60e0add-d345-4cb8-b796-eae0d497af96",
+ * |   "GLIB_DOMAIN" : "rauc-event",
+ * |   "RAUC_EVENT_TYPE" : "boot",
+ * |   "BOOT_ID" : "16655d2c-c5ca-48d3-bea8-7b95c803b4b2",
+ * |   "BUNDLE_HASH" : "unknown"
+ * | }
+ * | {
+ * |   "date": "2023-06-14T21:42:41Z"
+ * |   "MESSAGE" : "Marked slot rootfs.0 as active.",
+ * |   "MESSAGE_ID" : "8b5e7435-e105-4d86-8582-78e7544fe6da",
+ * |   "GLIB_DOMAIN" : "rauc-event",
+ * |   "RAUC_EVENT_TYPE" : "mark",
+ * |   "RAUC_SLOT" : "rootfs.0",
+ * |   "BUNDLE_HASH" : "",
+ * |   "SLOT_BOOTNAME" : "A"
+ * | }
+ *
+ * @param log_level log level, either from GLogLevelFlags, or a user-defined level
+ * @param fields key–value pairs of structured data forming the log message.
+ *               [array length=n_fields]
+ * @param n_fields number of elements in the fields array
+ * @param pretty Whether or not to pretty-print output (have human-readable line breaks)
+ *
+ * @return newly-allocated formatted string (without trailing newline)
+ */
+#if ENABLE_JSON
+static gchar *event_log_format_fields_json(GLogLevelFlags log_level,
+		const GLogField *fields, gsize n_fields, gboolean pretty)
+{
+	g_autoptr(GDateTime) now = NULL;
+	g_autofree gchar *now_formatted = NULL;
+	g_autoptr(JsonGenerator) gen = NULL;
+	g_autoptr(JsonNode) root = NULL;
+	g_autoptr(JsonBuilder) builder = json_builder_new();
+
+	now = g_date_time_new_now_utc();
+	now_formatted = g_date_time_format(now, "%Y-%m-%dT%H:%M:%SZ");
```

Comment:
Perhaps use `g_date_time_format_iso8601()`, as this also provides sub-second precision? That could be useful to correlate events if as lot is happening.

Reply:
I chose the format as this is consistent with the existing ones.

Another thing is that `g_date_time_format_iso8601()` requires glib 2.62 as a minimal version.

If we agree that it is worth the effort, we can either bump glib minimal version or conditionally import the method.

2.62 is available since Yocto/poky `dunfell`, Ubuntu 20.04, Debian oldstable (`bullseye`)

Reply:
As the current format is also ISO8601 (just without microsecond precision), we can use this for now and move to `g_date_time_format_iso8601()` later.